### PR TITLE
[SPARK-30269][SQL] Should use old partition stats to decide whether to update stats when analyzing partition

### PR DESCRIPTION
**Pull request by wzhfy**
### What changes were proposed in this pull request?
It's an obvious bug: currently when analyzing partition stats, we use old table stats to compare with newly computed stats to decide whether it should update stats or not.

### Why are the changes needed?
bug fix

### Does this PR introduce any user-facing change?
no

### How was this patch tested?
add new tests

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/StatisticsCollectionTestBase.scala [hunk 1]
```diff
@@ -26,6 +26,7 @@ import scala.collection.mutable
 import scala.util.Random
 
 import org.apache.spark.sql.catalyst.{QualifiedTableName, TableIdentifier}
+import org.apache.spark.sql.catalyst.catalog.CatalogTypes.TablePartitionSpec
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/StatisticsCollectionTestBase.scala]: Could you fix the import order to pass the GitHub Action and Jenkins?

Comment (wzhfy) [reply sql/core/src/test/scala/org/apache/spark/sql/StatisticsCollectionTestBase.scala]: forgot to fix this one. thanks for reminding!

**Review by HyukjinKwon — APPROVED** [review]

**Review by maropu — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master and branch-2.4.

**Comment by HyukjinKwon** [comment]
Oops, it has a conflict against branch-2.4. @wzhfy mind if I ask to open a backport PR?

**Comment by wzhfy** [comment]
@HyukjinKwon of course not, a backport would be great

**Comment by wzhfy** [comment]
@HyukjinKwon I create [pr#26963](https://github.com/apache/spark/pull/26963) and [pr#26964](https://github.com/apache/spark/pull/26964) for branch 2.4 and 2.3 respectively, could you take a look?

**Pull request closed without merging**